service/matching/task_queue_manager.go

```diff
@@ -371,12 +371,18 @@ func (c *taskQueueManagerImpl) SetInitializedError(err error) {
 	}
 }
 
-func (c *taskQueueManagerImpl) SetUserDataInitialFetch(err error) {
-	c.userDataInitialFetch.Set(struct{}{}, err)
-	if err != nil {
-		// We can't recover from here without starting over, so unload the whole task queue
-		c.lostOwnership.Store(true) // not really lost ownership but we want to skip the last write
-		c.unloadFromEngine()
+// Sets user data enabled/disabled and marks the future ready (if it's not ready yet).
+func (c *taskQueueManagerImpl) SetUserDataInitialFetch(disabledError error, futureError error) {
+	// Always set disabled state even if we're not setting the future.
+	c.db.disableUserData(disabledError)
+
+	if !c.userDataInitialFetch.Ready() {
+		c.userDataInitialFetch.Set(struct{}{}, futureError)
+		if futureError != nil {
+			// We can't recover from here without starting over, so unload the whole task queue
+			c.lostOwnership.Store(true) // not really lost ownership but we want to skip the last write
```

Comment:
The majority of places that use this value now have this same comment. I think the original meaning is lost, and we should just rename it to whatever it actually is for (which seems to be whether we should do a final write): 
<img width="1174" alt="image" src="https://github.com/temporalio/temporal/assets/5942963/ff7bde45-a4cd-4a5f-81d7-f6afcd2cef20">

Reply:
And a more meaningful comment would be not *what* this is doing (skipping the last write), but *why*. Why do we want to skip the last write here? I think we should document that.